crates/control/src/role_assignment.rs

```diff
@@ -740,85 +640,119 @@ fn pick_role_with_penalties(
     striker_player_number: PlayerNumber,
     optional_roles: &[Role],
 ) -> Role {
-    let mut role_assignment: Players<Option<Role>> = Players {
-        one: None,
-        two: None,
-        three: None,
-        four: None,
-        five: None,
-        six: None,
-        seven: None,
-    };
-
-    role_assignment[striker_player_number] = Some(Role::Striker);
-    let mut unassigned_robots = 6;
-
-    unassigned_robots -= penalties
+    let mut unassigned_players: VecDeque<_> = penalties
         .iter()
-        .filter(|(_player, &penalty)| penalty.is_some())
-        .count();
+        .filter_map(|(player_number, penalty)| {
+            (player_number != striker_player_number && penalty.is_none()).then_some(player_number)
+        })
+        .collect();
 
-    if unassigned_robots > 0 {
-        unassigned_robots =
-            assign_keeper_or_replacement_keeper(unassigned_robots, penalties, &mut role_assignment);
+    let mut role_assignment: Players<Option<Role>> = Players::new(None);
+    role_assignment[striker_player_number] = Some(Role::Striker);
+    if let Some(keeper) = unassigned_players.pop_front() {
+        role_assignment[keeper] = Some(match keeper {
+            PlayerNumber::One => Role::Keeper,
+            _ => Role::ReplacementKeeper,
+        })
     }
 
-    for &optional_role in optional_roles.iter().take(unassigned_robots) {
-        if needs_assignment(PlayerNumber::Two, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Two] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Three, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Three] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Four, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Four] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Five, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Five] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Six, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Six] = Some(optional_role);
-        } else if needs_assignment(PlayerNumber::Seven, penalties, &role_assignment) {
-            role_assignment[PlayerNumber::Seven] = Some(optional_role);
-        }
+    for (player_number, &optional_role) in unassigned_players.into_iter().zip(optional_roles) {
+        role_assignment[player_number] = Some(optional_role)
     }
 
     role_assignment[own_player_number].unwrap_or_default()
```

Comment:
Can we remove `Default` from `Role` entirely? I do not like to have `Default` for a role, as this depends on the context. There is no general, natural default value to a role. Being a striker is a implementation in some fallback cases of the role assignment.

The consequence of that is to explicitly state that we want to assign striker(s) in respective cases